### PR TITLE
Move TextAlign out of TextStyle

### PR DESCRIPTION
**Pull request by abarth**
TextAlign applies to a whole paragraph instead of applying to an individual
text span. This patch moves the property out of TextStyle and into a separate
property on Text and RichText.

### CODE REVIEW

File: packages/flutter/lib/src/rendering/child_view.dart [hunk 1]
```diff
@@ -287,8 +287,9 @@ class RenderChildView extends RenderBox {
       _child._setChildProperties(_physicalWidth, _physicalHeight, scale);
       assert(() {
         if (_view == null) {
-          _debugErrorMessage ??= new TextPainter()
-            ..text = new TextSpan(text: 'Child view are supported only when running in Mojo shell.');
+          _debugErrorMessage ??= new TextPainter(
+            text: new TextSpan(text: 'Child view are supported only when running in Mojo shell.')
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/child_view.dart]: views

Comment (Hixie) [inline packages/flutter/lib/src/rendering/child_view.dart]: or is

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 2]
```diff
@@ -23,13 +23,18 @@ class RenderParagraph extends RenderBox {
...
+    assert(value != null);
     _textPainter.text = value;
     markNeedsLayout();
   }
 
+  /// How the text should be aligned horizontally.
+  TextAlign get textAlign => _textPainter.textAlign;
+  void set textAlign(TextAlign value) {
+    _textPainter.textAlign = value;
+    markNeedsLayout();
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/paragraph.dart]: changing textalign should only require a repaint, no?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 3]
```diff
@@ -1931,13 +1942,41 @@ class DefaultTextStyle extends InheritedWidget {
...
+  /// Returned from [of] when the given [BuildContext] doesn't have an enclosing default text style.
+  const DefaultTextStyle.fallback() : style = const TextStyle(), textAlign = null;
+
+  /// Creates a default text style that inherits from the given [BuildContext].
+  ///
+  /// The given [style] is merged with the [style] from the default text style
+  /// for the given [BuildContext] and, if non-null, the given [textAlign]
+  /// replaces the [textAlign] from the default text style for the given
+  /// [BuildContext].
+  DefaultTextStyle.inherit({
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: if you made this a factory, you could do the .of() call just once.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 4]
```diff
@@ -1989,18 +2028,33 @@ class Text extends StatelessWidget {
...
-      return DefaultTextStyle.of(context).merge(style);
+      return DefaultTextStyle.of(context).style.merge(style);
     else
       return style;
   }
 
   @override
   Widget build(BuildContext context) {
+    TextStyle effectiveTextStyle = style;
+    DefaultTextStyle defaultTextStyle;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: for symmetry, swap these two around

**Comment by Hixie** [comment]
Needs tests.

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/lib/src/widgets/basic.dart (hunk 3 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/basic.dart]: Ah, good idea.

**Pull request merged**